test/podman-tests.sh

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+
+# Do not set -e as we want to work on all results of `rpm -q dnf[5]`.
+set -xo pipefail
+
+if [[ $1 == '' ]]; then
+    echo -e "Usage: podman-tests.sh TEST_TYPE STREAM\nTEST_TYPE can be 'e2e' or 'system'\nSTREAM can be 'upstream' or 'downstream'"
+    exit 0
+fi
+
+# `rhel` macro exists on RHEL, CentOS Stream, and Fedora ELN
+# `centos` macro exists only on CentOS Stream
+CENTOS_VERSION=$(rpm --eval '%{?centos}')
+RHEL_VERSION=$(rpm --eval '%{?rhel}')
+
+# For upstream tests, we need to test with podman and other packages from the
+# podman-next copr. For downstream tests (bodhi, errata), we don't need any
+# additional setup
+if [[ $2 == "upstream" ]]; then
+    # Use CentOS Stream 10 copr target for RHEL-10 until EPEL 10 becomes
+    # available
+    if [[ -n $CENTOS_VERSION || $RHEL_VERSION -ge 10 ]]; then
+        dnf -y copr enable rhcontainerbot/podman-next centos-stream-$CENTOS_VERSION
+    else
+        dnf -y copr enable rhcontainerbot/podman-next
+    fi
+    echo "priority=5" >> /etc/yum.repos.d/_copr:copr.fedorainfracloud.org:rhcontainerbot:podman-next.repo
+fi
+
+# Remove testing-farm repos if they exist
+rm -f /etc/yum.repos.d/tag-repository.repo
+
+# Install common dependencies for tests
+dnf -y install podman
+
+cat /etc/redhat-release
+rpm -q container-selinux podman
+
+if [[ $1 == "e2e" ]]; then
+    # Install dependencies for tests
+    dnf -y install golang
+    rpm -q golang
+
+    # /tmp is often unsufficient
+    export TMPDIR=/var/tmp
+
+    # dnf5 contains breaking changes
+    # Either of `dnf` OR `dnf5` will be installed, never both.
+    # To fetch srpm, dnf uses `--source`, dnf5 uses `--srpm`.
+    rpm -q dnf5
+    if [[ $? -eq 0 ]]; then
+        SRPM_OPTS="--srpm"
+    else
+        SRPM_OPTS="--source"
+    fi
+
+    # Fetch and extract latest podman source from the highest priority dnf repo
+    # NOTE: On upstream pull-requests, the srpm will be fetched from the
+    # podman-next copr while on bodhi updates, it will be fetched from Fedora's
+    # official repos.
+    dnf download $SRPM_OPTS podman
+    rpm2cpio podman*.src.rpm | cpio -di
+    tar zxf podman-*.tar.gz
+
+    # Run podman e2e tests
+    cd podman-*/test/e2e
+    PODMAN_BINARY=/usr/bin/podman go test -v config.go config_amd64.go common_test.go libpod_suite_test.go run_selinux_test.go
+fi
+
+if [[ $1 == "system" ]]; then
+    # Enable EPEL to fetch bats
+    if [[ -n $(rpm --eval '%{?rhel}') ]]; then
+        # Until EPEL 10 is available use epel-9 for all RHEL and CentOS Stream
+        dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm
+        sed -i 's/$releasever/9/g' /etc/yum.repos.d/epel.repo
+    fi
```

Comment:
This can be controlled with [`feature`](https://tmt.readthedocs.io/en/stable/spec/plans.html#feature), but I think it's by default on `testing-farm`, not sure about RHEL though. You can alter the order of `prepare` to make sure that only `bats` is downloaded from epel.

Reply:
i've had tons of trouble making epel work on internal testing farm. So, I was looking for something that'd just work across all testing farms. IIRC, bats rpm pulls in parallel which also ends up pulling in a bunch of perl packages. So, I wonder if `only bats` would work.

I would've preferred to just install bats from upstream using their install script, but some people in the team had a preference for rpms.

Reply:
I mean not `only bats`, but rather pulling in at the appropriate time. But this was in the context of having multiple `prepare` steps. For the context of reproducibility, this would be preferred indeed.